plutus-ledger/src/Ledger/Tx.hs

```diff
@@ -75,99 +73,119 @@ import Data.Proxy (Proxy (Proxy))
-      -- public key hash.
-      _ciTxOutAddress         :: Address,
+-- | Transaction output represented off-chain
+-- add more
+data OffChainTxOut =
+    PublicKeyOffChainTxOut' {
+      -- | Hash of the public key protecting the transaction output.
+      ocTxOutAddress            :: V1.PubKeyHash,
       -- | Value of the transaction output.
-      _ciTxOutValue           :: V1.Value,
-      -- | Optional datum attached to the transaction output.
-      _ciTxOutPublicKeyDatum  :: V2.OutputDatum,
+      ocTxOutValue              :: V1.Value,
+      -- | Optional datum hash attached to the transaction output.
+      ocTxOutPublicKeyDatumHash :: Maybe V1.DatumHash,
+      -- | Optional datum attached to the transaction output, either found
+      -- inline or resolved from its hash offchain.
+      ocTxOutDatum              :: Maybe V1.Datum,
       -- | Optional reference script attached to the transaction output.
-      _ciTxOutReferenceScript :: Maybe V1.Script
+      ocTxOutReferenceScript    :: Maybe V1.Script
     }
-  | ScriptChainIndexTxOut {
-      -- | Address of the transaction output. The address is protected by a
-      -- script.
-      _ciTxOutAddress         :: Address,
+  | ScriptOffChainTxOut' {
+      -- | Hash of the validator protecting the transaction output.
+      ocTxOutValidatorHash   :: V1.ValidatorHash,
       -- | Value of the transaction output.
-      _ciTxOutValue           :: V1.Value,
-      -- | Datum attached to the transaction output, either in full or as a
-      -- hash reference. A transaction output protected by a Plutus script
-      -- is guardateed to have an associated datum.
-      _ciTxOutScriptDatum     :: Either V1.DatumHash V1.Datum,
+      ocTxOutValue           :: V1.Value,
+      -- | FIXME Hash of the datum attached to the transaction output.
+      ocTxOutScriptDatumHash :: V1.DatumHash,
+      -- | Validator script protecting the transaction output, resolved
+      -- from its hash off-chain
+      ocTxOutValidator       :: Maybe V1.Validator,
+      -- | Datum attached to the transaction output, either found inline or
+      -- resolved from its hash off-chain. A transaction output protected
+      -- by a Plutus script is need to have an associated datum to be
+      -- spendable.
+      ocTxOutDatum           :: Maybe V1.Datum,
       -- | Optional reference script attached to the transaction output.
       -- The reference script is, in genereal, unrelated to the validator
       -- script althought it could also be the same.
-      _ciTxOutReferenceScript :: Maybe V1.Script,
-      -- | Validator protecting the transaction output, either in full or
-      -- as a hash reference.
-      _ciTxOutValidator       :: Either V1.ValidatorHash V1.Validator
+      ocTxOutReferenceScript :: Maybe V1.Script
     }
   deriving (Show, Eq, Serialise, Generic, ToJSON, FromJSON, OpenApi.ToSchema)
 
-makeLenses ''ChainIndexTxOut
-makePrisms ''ChainIndexTxOut
```

Comment:
Should we bring those? We use `lens` in our codebase, so having them seems useful.

Reply:
Let's discuss in chat.